File: app/analytics/session_analyzer.py
from __future__ import annotations

from datetime import datetime

import numpy as np

from .metrics import AnalysisConfig, analyze_effort_samples
from .models import EffortMetrics, ForceSample, SessionSummary
from .segmentation import SegmenterConfig, segment_efforts


def analyze_session(
    samples: list[ForceSample],
    hand: str,
    segment_cfg: SegmenterConfig,
    analysis_cfg: AnalysisConfig,
    session_id: str,
    started_at_iso: str,
    ended_at_iso: str | None = None,
) -> tuple[list[EffortMetrics], SessionSummary]:
    if ended_at_iso is None:
        ended_at_iso = datetime.utcnow().isoformat(timespec="seconds")

    ranges = segment_efforts(samples, segment_cfg)
    efforts: list[EffortMetrics] = []

    for i, (start_idx, end_idx) in enumerate(ranges, start=1):
        seg = samples[start_idx : end_idx + 1]
        if len(seg) < 2:
            continue
        efforts.append(analyze_effort_samples(seg, effort_id=i, config=analysis_cfg))

    peaks = np.array([e.peak_total_kg for e in efforts], dtype=float)
    if len(peaks) >= 2:
        x = np.arange(len(peaks), dtype=float)
        fatigue_slope = float(np.polyfit(x, peaks, 1)[0])
        first_to_last_drop_pct = float(((peaks[-1] - peaks[0]) / peaks[0]) * 100.0) if peaks[0] > 1e-9 else 0.0
    else:
        fatigue_slope = 0.0
        first_to_last_drop_pct = 0.0

    summary = SessionSummary(
        session_id=session_id,
        started_at_iso=started_at_iso,
        ended_at_iso=ended_at_iso,
        hand=hand,
        efforts_count=len(efforts),
        best_peak_kg=float(np.max(peaks)) if len(peaks) else 0.0,
        avg_peak_kg=float(np.mean(peaks)) if len(peaks) else 0.0,
        fatigue_slope_kg_per_effort=fatigue_slope,
        first_to_last_drop_pct=first_to_last_drop_pct,
    )
    return efforts, summary

File: packages/core/src/sessionAnalyzer.ts
import type { EffortMetrics, ForceSample, SessionSummary, Hand } from './types.ts';
import type { AnalysisConfig } from './metrics.ts';
import type { SegmenterConfig } from './segmentation.ts';
import { analyzeEffortSamples } from './metrics.ts';
import { segmentEfforts } from './segmentation.ts';

function linearRegressionSlope(values: number[]): number {
  if (values.length < 2) return 0;
  const count = values.length;
  let sumX = 0;
  let sumY = 0;
  let sumXY = 0;
  let sumX2 = 0;

  for (let i = 0; i < count; i++) {
    sumX += i;
    sumY += values[i];
    sumXY += i * values[i];
    sumX2 += i * i;
  }

  const denominator = count * sumX2 - sumX * sumX;
  if (Math.abs(denominator) < 1e-12) return 0;
  return (count * sumXY - sumX * sumY) / denominator;
}

export function analyzeSession(
  samples: ForceSample[],
  hand: Hand,
  segmentCfg: SegmenterConfig,
  analysisCfg: AnalysisConfig,
  sessionId: string,
  startedAtIso: string,
  endedAtIso?: string,
): { efforts: EffortMetrics[]; summary: SessionSummary } {
  const ended = endedAtIso ?? new Date().toISOString();
  const ranges = segmentEfforts(samples, segmentCfg);
  const efforts: EffortMetrics[] = [];

  for (let i = 0; i < ranges.length; i++) {
    const [startIdx, endIdx] = ranges[i];
    const segment = samples.slice(startIdx, endIdx + 1);
    if (segment.length < 2) continue;
    efforts.push(analyzeEffortSamples(segment, i + 1, analysisCfg));
  }

  const peaks = efforts.map(effort => effort.peakTotalKg);
  const fatigueSlopeKgPerEffort = linearRegressionSlope(peaks);

  let bestPeakKg = 0;
  let avgPeakKg = 0;
  let firstToLastDropPct = 0;
  if (peaks.length > 0) {
    bestPeakKg = Math.max(...peaks);
    avgPeakKg = peaks.reduce((sum, value) => sum + value, 0) / peaks.length;
    if (peaks.length >= 2 && peaks[0] > 1e-9) {
      firstToLastDropPct = ((peaks[peaks.length - 1] - peaks[0]) / peaks[0]) * 100;
    }
  }

  return {
    efforts,
    summary: {
      sessionId,
      startedAtIso,
      endedAtIso: ended,
      hand,
      effortsCount: efforts.length,
      bestPeakKg,
      avgPeakKg,
      fatigueSlopeKgPerEffort,
      firstToLastDropPct,
    },
  };
}
